Sources/TuistCore/Graph/Graph.swift

```diff
@@ -24,28 +24,65 @@ enum GraphError: FatalError {
 public class Graph: Encodable {
     // MARK: - Attributes
 
-    private let cache: GraphLoaderCaching
+    /// Name of the graph.
     public let name: String
+
+    /// Path to the directory where the graph has been loaded from.
     public let entryPath: AbsolutePath
+
+    /// The entry nodes of the graph.
     public let entryNodes: [GraphNode]
 
+    /// Dictionary whose keys are paths to directories where projects are defined, and the values are the representation of the projects.
+    public let projects: WeakArray<Project>
+
+    /// Dictionary whose keys are paths to directories where projects are defined, and the values are the CocoaPods nodes define in them.
+    /// If none of the nodes of the graph references a CocoaPods node, the node gets released from memory.
+    public let cocoapods: WeakArray<CocoaPodsNode>
+
+    /// Dictionary whose keys are path to directories where projects are defined, and the values are packages defined in that project.
+    /// If none of the nodes of the graph references a Package node, the node gets released from memory.
+    public let packages: WeakArray<PackageNode>
+
+    /// Dictionary whose keys are path to directories where projects are defined, and the values are precompiled nodes defined in them.
+    /// If none of the nodds references a precompiled node, the node gets released from memory.
+    public let precompiled: WeakArray<PrecompiledNode>
+
+    /// Returns all the frameorks that are part of the graph.
+    public var frameworks: [FrameworkNode] { precompiled.compactMap { $0 as? FrameworkNode } }
+
+    /// Dictionary whose keys are path to directories where projects are defined, and the values are target nodes defined in them.
+    public let targets: [AbsolutePath: WeakArray<TargetNode>]
```

Comment:
Not sure I follow the reasoning behind weak arrays :/ - the graph loader would always load referenced nodes, and if any need to be removed that can be done separately ahead of initializing the graph. 

I worry this may make it difficult to move the graph/models towards being a value types. Perhaps worth separating this to a different PR.

Reply:
Ah I think I get the reasoning - it's to remove orphaned nodes after mapping graphs to swap out dependencies? (https://github.com/tuist/tuist/pull/1056/) 

There may be an alternate way we could solve that problem without resorting to weak arrays :)  - I'll have a think 💭

Reply:
Exactly. The only strong reference from the the entrypoints and the Target.dependencies attribute. Thanks to the WeakArrays, when we delete nodes for the graph, the references from Graph are orphaned.

Reply:
I’m a bit skeptical regarding moving the graph to value types. Accessing nodes through references between the class instance makes traversing operations fast. Moreover, if we want to do mutations with a value-typed graph, we’ll have to traverse the graph after each mutation, to ensure all elements of it are consistent.